src/main/scala/com/databricks/spark/sql/perf/mllib/classification/NaiveBayes.scala

```diff
@@ -0,0 +1,65 @@
+package com.databricks.spark.sql.perf.mllib.classification
+
+import org.apache.spark.ml
+import org.apache.spark.ml.{ModelBuilder, PipelineStage, Transformer}
+import org.apache.spark.ml.evaluation.{Evaluator, MulticlassClassificationEvaluator}
+import org.apache.spark.ml.linalg.{DenseMatrix, Vectors}
+
+import com.databricks.spark.sql.perf.mllib.OptionImplicits._
+import com.databricks.spark.sql.perf.mllib._
+import com.databricks.spark.sql.perf.mllib.data.DataGenerator
+
+/** Object containing methods used in performance tests for (multinomial) NaiveBayesModels */
+object NaiveBayes extends BenchmarkAlgorithm
+  with TestFromTraining with TrainingSetFromTransformer with ScoringWithEvaluator {
+
+  override protected def initialData(ctx: MLBenchContext) = {
+    import ctx.params._
+    val rng = ctx.newGenerator()
+    // Max possible arity of a feature in generated training/test data for NaiveBayes models
+    val maxFeatureArity = 20
+    // All features for Naive Bayes must be categorical, i.e. have arity >= 2
+    val featureArity = 0.until(numFeatures).map(_ => 2 + rng.nextInt(maxFeatureArity - 2)).toArray
+    val df = DataGenerator.generateMixedFeatures(
+      ctx.sqlContext,
+      numExamples,
+      ctx.seed(),
+      numPartitions,
+      featureArity)
+    df
+  }
+
+  override protected def trueModel(ctx: MLBenchContext): Transformer = {
+    import ctx.params._
+    val rng = ctx.newGenerator()
+    // pi = log of class priors, whose dimension is C (number of classes)
+    // theta = log of class conditional probabilities, whose dimension is C (number of classes)
+    // by D (number of features)
+    val unnormalizedProbs = 0.until(numClasses).map(_ => rng.nextDouble() + 1e-5).toArray
+    val piArray = unnormalizedProbs.map(prob => math.log(prob / unnormalizedProbs.sum))
```

Comment:
You can also write it as log(prob) - log(sum) to make it more numerically stable.